sylt-compiler/src/typechecker.rs

```diff
@@ -791,6 +1006,86 @@ impl TypeChecker {
         }
     }
 
+    fn is_indexed_by(
+        &mut self,
+        span: Span,
+        ctx: TypeCtx,
+        a: usize,
+        b: usize,
+    ) -> Result<(), Vec<Error>> {
+        match (self.find_type(a), self.find_type(b)) {
+            (Type::Unknown, _) => Ok(()),
+            (_, Type::Unknown) => Ok(()),
+
+            (Type::List(_), Type::Int) => Ok(()),
+            (Type::Tuple(_), Type::Int) => Ok(()),
+            // TODO(ed): Sets!
+            (Type::Dict(k, _), _) => {
+                self.unify(span, ctx, k, b)?;
+                Ok(())
+            }
+
+            _ => {
+                return err_type_error!(
+                    self,
+                    span,
+                    ctx,
+                    TypeError::BinOp {
+                        lhs: self.bake_type(a),
+                        rhs: self.bake_type(b),
+                        op: "Indexing".to_string(),
+                    }
+                )
+            }
+        }
+    }
+
+    fn is_given_by_indexing(
+        &mut self,
+        span: Span,
+        ctx: TypeCtx,
+        a: usize,
+        b: usize,
+    ) -> Result<(), Vec<Error>> {
+        match self.find_type(a) {
+            Type::Unknown => Ok(()),
+
+            Type::Tuple(_) => {
+                // NOTE(ed): If we get here - it means we're checking the constraint, but the
+                // constraint shouldn't be added because we only ever check constants.
+                return err_type_error!(
+                    self,
+                    span,
+                    ctx,
+                    TypeError::Exotic,
+                    "Tuples can only be indexed by integer constants"
+                );
+            }
```

Comment:
This is actually wrong - we need to make sure the integer is constants - we need a marker that we can track throughout the codebase, or we will allow invalid programs when passing tuples to functions! D: